### ISSUE TITLE
Impossible to store a struct field of type serde_json::StreamDeserializer<?, T>

### ISSUE THREAD

**Issue by dflemstr** [serde-rs/json#259]
While working on https://github.com/dflemstr/rq/issues/151 I noticed that `serde_json::Deserializer::from_reader(r).into_iter()` returns a Voldemort type, so it can't be stored in a struct.  I want to create a:

```rust
pub struct JsonSource<R>(serde_json::StreamDeserializer<R, value::Value>)
    where R: serde_json::de::Read;
```

...but `R` is not actually a variable, but rather the private type `IteratorRead<Bytes<R2>>` for a type `R2: std::io::Read`.  The `IteratorRead` struct would have to be public for the type to be nameable.

**Comment by dtolnay** [comment]
I just fixed this earlier today in serde_json 0.9.7. Your code works for me:

```rust
extern crate serde_json;
use serde_json::value;

pub struct JsonSource<R>(serde_json::StreamDeserializer<R, value::Value>)
    where R: serde_json::de::Read;

fn main() {}
```

**Comment by dflemstr** [comment]
Try instantiating the struct with a `serde_json::Deserializer::from_reader(r).into_iter()` in it and you will see what I mean!

**Comment by dtolnay** [comment]
Can you share your code? I'm still not getting it.

```rust
pub fn f() {
    let read = File::open("/etc/passwd").unwrap();
    let _ = JsonSource(serde_json::Deserializer::from_reader(read).into_iter());
}
```

**Comment by dflemstr** [comment]
Sorry for not creating a more complete example at first, I thought the description would be enough.  Basically, in your example you avoid naming the type since you don't specify the type of the `_`.  But this is my situation:

```rust
extern crate serde_json;

use std::io;

use serde_json::value;

pub struct JsonSource<R>(serde_json::StreamDeserializer<R, value::Value>)
    where R: serde_json::de::Read;

fn source<IoRead>(r: R) -> JsonSource<???>
    where IoRead: io::Read
{
    serde_json::Deserializer::from_reader(r).into_iter()
}
```

What is `???`?

**Comment by dtolnay** [comment]
You're right this is still bad. I will expose all the types.

**Comment by dtolnay** [comment]
Addressed in #260. I'll cut a release tomorrow morning.

**Comment by dtolnay** [comment]
Released in serde_json 0.9.8.
